src/main/scala/org/allenai/pipeline/Pipeline.scala

```diff
@@ -24,30 +28,56 @@ import java.util.Date
 trait Pipeline extends Logging {
   def artifactFactory: FlatArtifactFactory[String] with StructuredArtifactFactory[String]
 
-  def run(rawTitle: String, outputs: Producer[_]*): Unit = {
-    try {
-      val outputInfo = outputs.map(_.stepInfo)
-      require(
-        outputInfo.forall(_.outputLocation.isDefined),
-        s"Running a pipeline without persisting the output: ${
-          outputInfo.filter(_.outputLocation.isEmpty)
-            .map(_.className).mkString(",")
-        }"
-      )
+  private[this] val persistedSteps: ListBuffer[Producer[_]] = ListBuffer()
+
+  /** Run the pipeline.  All steps that have been persisted will be computed, along with any upstream dependencies */
+  def run(title: String): Unit = {
+    run(title, persistedSteps.toSeq)
+  }
 
+  /** Run only specified steps in the pipeline.  Upstream dependencies must exist already.  They will not be computed */
+  def runOnly(title: String, runOnlyTargets: Producer[_]*) {
+    val targets = runOnlyTargets.flatMap(s => persistedSteps.find(_.stepInfo.signature == s.stepInfo.signature))
+    require(targets.size == runOnlyTargets.size, "Specified targets are not members of this pipeline")
+
+    val persistedStepsInfo = persistedSteps.map(_.stepInfo).toSet
+    val overridenStepsInfo = targets.map(_.stepInfo).toSet
+    val nonPersistedTargets = overridenStepsInfo -- persistedStepsInfo
+    require(
+      nonPersistedTargets.size == 0,
+      s"Running a pipeline without persisting the output: [${nonPersistedTargets.map(_.className).mkString(",")}]"
+    )
+    val allDependencies = targets.flatMap(Workflow.upstreamDependencies)
+    val nonExistentDependencies =
+      for {
+        p <- allDependencies if p.isInstanceOf[PersistedProducer[_, _ <: Artifact]]
+        pp = p.asInstanceOf[PersistedProducer[_, _ <: Artifact]]
+        if !overridenStepsInfo(pp.stepInfo)
+        if !pp.artifact.exists
+      } yield pp.stepInfo
+    require(nonExistentDependencies.size == 0, {
+      val targetNames = overridenStepsInfo.map(_.className).mkString(",")
+      val dependencyNames = nonExistentDependencies.map(_.className).mkString(",")
```

Comment:
I'd move towards using the "persist" name instead of class name. class names are confusing, as they can be reused in several places.